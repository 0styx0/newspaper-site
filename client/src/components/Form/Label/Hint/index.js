import React from 'react';
import PropTypes from 'prop-types';


export default function Hint(props) {


    return <span>
               <abbr onClick={props.onClick} title={props.title}>?</abbr>
               {props.revealHint ? <div className="abbrMessage">{props.title}</div> : <span />}
           </span>
}

Hint.propTypes = {
    onClick: PropTypes.func.isRequired,
    title: PropTypes.string.isRequired, // what will be shown when revealHint is true
    revealHint: PropTypes.bool // show or not to show, that is the question
}