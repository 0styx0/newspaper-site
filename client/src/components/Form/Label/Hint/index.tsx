import * as React from 'react';

interface Props {
    onClick: Function;
    title: string; // what will be shown when revealHint is true
    revealHint: boolean; // show or not to show, that is the question
    children: JSX.Element;
}

/**
 * Appends a hint to props.children
 */
export default function Hint(props: Props) {

    return React.cloneElement(props.children, {
        children: props.children.props.children.concat([
            (
                <span>
                  <abbr onClick={props.onClick as any} title={props.title}>?</abbr>
                  {props.revealHint ? <div className="abbrMessage">{props.title}</div> : <span />}
                </span>
                )
        ])
    });
}
