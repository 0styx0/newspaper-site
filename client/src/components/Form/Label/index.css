label {
  font-weight: bold;
  display: block;
  margin-top: 5px;
  text-align: right;
}

label abbr {
    padding-left: 3px;
    color: blue;
    font-size: 9pt;
    text-decoration: none;
}

.abbrMessage {
    text-align: left;
    font-size: 10pt;
    background-color: white;
    max-width: 20em;
    margin: 2px 5px;
    position: relative;
    left: 4.5em;
}
