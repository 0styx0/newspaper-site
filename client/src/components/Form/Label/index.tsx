import * as React from 'react';

import HintContainer from './Hint/container';

import './index.css';

interface Props {
    value: string;
    required?: boolean;
    abbr?: string;
    children: JSX.Element;
}

/**
 * @prop label, children, abbr, required @see Select
 *
 * @return html label as parent to props.children with text of prop.value concatted with ': ',
   if abbr is given, that's put in too, and if required is given, a red asterisk is put in too
 */
export default function Label(props: Props) {

    const label = props.value + ": ";

    return (
        <label>{label}
            {props.children}
            {!!props.required ? <span className="danger">*</span> : ""}
            {!!props.abbr ? <HintContainer title={props.abbr} /> : ""}
        </label>
    )
}
