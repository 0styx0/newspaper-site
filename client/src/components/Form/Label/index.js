import React from 'react';
import Hint from './Hint';

import './index.css';

/**
 * @prop label, children, abbr, required @see Select
 *
 * @return html label as parent to props.children with text of prop.value concatted with ': ',
   if abbr is given, that's put in too, and if required is given, a red asterisk is put in too
 */
export default function Label(props) {

    const label = props.value + ": ";

    return (
        <label>{label}
            {props.children}
            {!!props.required ? <span className="danger">*</span> : ""}
            {!!props.abbr ? <Hint title={props.abbr} /> : ""}
        </label>
    )
}

