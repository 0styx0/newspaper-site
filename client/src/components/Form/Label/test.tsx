import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Label from './';

describe('<Label>', () => {

    describe('snapshots', () => {

        /**
         * Takes snapshots
         */
        function snap(abbr?: string, required: boolean = false) {

            const tree = renderer.create(
                <Label
                    value="Label"
                    required={required}
                    abbr={abbr}
                    children={<input type="text" value="hello" />}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();

        }

        it('renders correctly when abbr is given and not required', () => snap('A very good hint'));

        it('renders correctly when abbr is NOT given and not required', () => snap());

        it('when required and no abbr', () => snap(undefined, true));

        it('when required and yes abbr', () => snap('Hello', true));
    });
});
