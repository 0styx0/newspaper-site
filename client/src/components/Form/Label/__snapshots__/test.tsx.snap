// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label> snapshots renders correctly when abbr is NOT given and not required 1`] = `
<label>
  Label: 
  <input
    type="text"
    value="hello"
  />
  
</label>
`;

exports[`<Label> snapshots renders correctly when abbr is given and not required 1`] = `
<label
  onInput={[Function]}
>
  Label: 
  <input
    type="text"
    value="hello"
  />
  
  <span>
    <abbr
      onClick={[Function]}
      title="A very good hint"
    >
      ?
    </abbr>
    <span />
  </span>
</label>
`;

exports[`<Label> snapshots when required and no abbr 1`] = `
<label>
  Label: 
  <input
    type="text"
    value="hello"
  />
  <span
    className="danger"
  >
    *
  </span>
</label>
`;

exports[`<Label> snapshots when required and yes abbr 1`] = `
<label
  onInput={[Function]}
>
  Label: 
  <input
    type="text"
    value="hello"
  />
  <span
    className="danger"
  >
    *
  </span>
  <span>
    <abbr
      onClick={[Function]}
      title="Hello"
    >
      ?
    </abbr>
    <span />
  </span>
</label>
`;
